refactor: wrap CCoinsViewCursor in unique_ptr

jamesob opened the pull request:
I tripped over this one for a few hours at the beginning of the week, so I've sort of got a personal vendetta against `CCoinsView::Cursor()` returning a raw pointer.

Specifically in the case of CCoinsViewDB, if a raw cursor is allocated and not freed, a cryptic leveldb assertion failure occurs on CCoinsViewDB destruction (`Assertion 'dummy_versions_.next_ == &dummy_versions_' failed.`).

This is a pretty simple change.

Related to: https://github.com/bitcoin/bitcoin/issues/21766
See also: https://github.com/google/leveldb/issues/142#issuecomment-414418135

jamesob committed:
refactor: wrap CCoinsViewCursor in unique_ptr

Specifically with CCoinsViewDB, if a raw cursor is allocated and
not freed, a cryptic leveldb assertion failure occurs on
CCoinsViewDB destruction.

See: https://github.com/google/leveldb/issues/142#issuecomment-414418135

MarcoFalke commented on src/txdb.cpp:
```diff
@@ -168,9 +168,10 @@ bool CBlockTreeDB::ReadLastBlockFile(int &nFile) {
+std::unique_ptr<CCoinsViewCursor> CCoinsViewDB::Cursor() const
 {
-    CCoinsViewDBCursor *i = new CCoinsViewDBCursor(const_cast<CDBWrapper&>(*m_db).NewIterator(), GetBestBlock());
+    std::unique_ptr<CCoinsViewDBCursor> i{
+        new CCoinsViewDBCursor(const_cast<CDBWrapper&>(*m_db).NewIterator(), GetBestBlock())};
```
```suggestion
    auto i{std::make_unique<CCoinsViewDBCursor>(const_cast<CDBWrapper&>(*m_db).NewIterator(), GetBestBlock())};
```

Can be written shorter with make_unique?

jamesob replied:
Good call!

maflcko reviewed (commented):
Concept ACK

practicalswift commented:
Concept ACK: nice catch!

Non-blocking nit: The list[1] of naked `new`:s in our code base is fairly short nowadays so it would be nice to avoid adding an unnecessary `new` to that list. Avoided by using `std::make_unique` as suggested by MarcoFalke :)

[1]
```
$ git grep -E 'new [A-Za-z0-9_]+\(' -- "*.cpp" "*.h" ":(exclude)src/qt/" ":(exclude)src/test/" \
      ":(exclude)src/bench/" ":(exclude)src/leveldb/"
```

jnewbery reviewed (commented):
Concept ACK. I agree with @MarcoFalke that we should use the standard `std::make_unique()` function when creating new unique pointers.

jamesob commented on src/test/fuzz/coins_view.cpp:
```diff
@@ -183,8 +183,8 @@ FUZZ_TARGET_INIT(coins_view, initialize_coins_view)
     {
-        const CCoinsViewCursor* coins_view_cursor = backend_coins_view.Cursor();
-        assert(coins_view_cursor == nullptr);
+        const std::unique_ptr<CCoinsViewCursor> coins_view_cursor = backend_coins_view.Cursor();
```
Good point, thanks.

jamesob commented:
Thanks for the quick looks everyone. I've made both changes suggested, though it's worth noting that I had to make the `CCoinsViewDBCursor` constructor public (from private) in order to allow `std::make_unique` to compile.

jnewbery commented:
Code review ACK 615c1adfb07b9b466173166dc2e53ace540e4b32

jonatack commented on src/txdb.h:
```diff
@@ -74,6 +74,8 @@ class CCoinsViewDB final : public CCoinsView
 class CCoinsViewDBCursor: public CCoinsViewCursor
 {
 public:
+    CCoinsViewDBCursor(CDBIterator* pcursorIn, const uint256 &hashBlockIn):
```
Verified the ctor needs to be public to not see `error: calling a private constructor of class 'CCoinsViewDBCursor'` when building. Could clang-format while touching it:
```suggestion
    CCoinsViewDBCursor(CDBIterator* pcursorIn, const uint256& hashBlockIn):
```

jonatack commented on src/test/fuzz/coins_view.cpp:
```diff
@@ -183,8 +183,8 @@ FUZZ_TARGET_INIT(coins_view, initialize_coins_view)
-        const CCoinsViewCursor* coins_view_cursor = backend_coins_view.Cursor();
-        assert(coins_view_cursor == nullptr);
+        std::unique_ptr<CCoinsViewCursor> coins_view_cursor = backend_coins_view.Cursor();
+        assert(!coins_view_cursor);
```
The change in this line seems to be style only.

jonatack reviewed (commented):
ACK 615c1adfb07b9b466173166dc2e53ace540e4b32

MarcoFalke commented on src/txdb.h:
```diff
@@ -74,6 +74,8 @@ class CCoinsViewDB final : public CCoinsView
 class CCoinsViewDBCursor: public CCoinsViewCursor
 {
 public:
+    CCoinsViewDBCursor(CDBIterator* pcursorIn, const uint256 &hashBlockIn):
```
Is there a reason to put this into the header? Previously the constructor was private, thus disallowing "illegal" access. Now, with it being public, the whole class could be moved to the cpp file?

jamesob replied:
Good call! No reason this needs to be exposed; moved to the .cpp file.

MarcoFalke commented on src/txdb.h:
```diff
@@ -84,8 +86,6 @@ class CCoinsViewDBCursor: public CCoinsViewCursor
-        CCoinsViewCursor(hashBlockIn), pcursor(pcursorIn) {}
     std::unique_ptr<CDBIterator> pcursor;
     std::pair<char, COutPoint> keyTmp;
 
```
line 92 still needed?

jamesob replied:
Yes, for e.g. `pcursor` access in `CCoinsViewDB::Cursor()`.

ryanofsky reviewed (approved):
Code review ACK 615c1adfb07b9b466173166dc2e53ace540e4b32. Changes all look correct and should be safer.

I am confused by this comment though: https://github.com/bitcoin/bitcoin/pull/22263#issuecomment-863256211

> it's worth noting that I had to make the CCoinsViewDBCursor constructor public (from private) in order to allow std::make_unique to compile.

I'm not sure if you think it is better or worse for the constructor to be public. If it is unsafe or bad to use to use the constructor, it would be nice to either:

- Document the new public constructor saying when it should or shouldn't be called
- Or keep the constructor private and just use `unique_ptr<CCoinsViewDBCursor>(new CCoinsViewDBCursor...)` with a comment about why a private constructor needs to be called.

These are basically just the two recommendations from https://abseil.io/tips/134#recommendations

jamesob committed:
move-only(ish): don't expose CCoinsViewDBCursor

No need for this to be a part of the header anymore.

Includes a small reference type style change.

jamesob committed:
doc: add comment about CCoinsViewDBCursor constructor

jamesob commented:
Thanks for the feedback all. I've added doc, fixed the reference type style, and made the `CCoinsViewDBCursor` private to the unit.

maflcko reviewed (approved):
review ACK 7ad414f4bfa74595ee5726e66f3527045c02a977 🔎

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 7ad414f4bfa74595ee5726e66f3527045c02a977 🔎
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiK0Av/YpJeCqfCFF8r/xKrmy/2fsY3yRopzMKExwCG6ZPVfJ09R0xfpAtoezgw
93OlloYTOppLdwCadZUt3tkY/Dd4YyHDH8mtlJif9B2NZtHQPRMkvy0QbB7I+XCc
O6JcW/tQi5FYsbXKUDLCzKqLC0VVJdhqijybnst0RQDXAYXHjKXwUZX1Tlywj4PC
5nPGFbQH5VbM9y54LgqjhBTogQP88N9Zn2T7jkm9vPgCX/FBhrnTqXHvIMssyjS1
cS6DUUxpF7iePYI4PugliofOl8bmlW7AIQcRj83qk8JCH9Bdr1v/WMC69kBGm06v
UpzsEJ0F9ZbmgHhvzT7Nkdgtq2QNu3R90HQlnVDd6m3qXSaSGUmZfrwCl67neHdw
BUlF5dQoXBgHdn+yms3iElNFxpfnwTRMzyWcBAjFJTGNQuzVIuHBSDRgJpjtTi67
5Wjay1OlBZNRkuej28hfTDD0UWqsxitPo1Me7C25c39WSeaN3wu5tw5hICkO4IIK
Pp60bF6D
=VU9g
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `6d7187a8c395c0137963dbd7752b6ff666a4fce6cdbe0e710567344055391df2  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401086d7187a8c395c0137963dbd7752b6ff666a4fce6cdbe0e710567344055391df2f0104ef5b7fd4ae54f454782e531f2efc88b08fff010396fb2036b11a4ef01af49bea0afd66f08f10460ccfc8ff00839cfb3f3cc1be6010083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010970df15c300e68282e5f77603844e74208f10460ccfc8ff00882c6405f75e5deae0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010ac09ac0ccc0dd3380986aa44d904290b08f0208968084ce1698e8df29f7f812b628f61d86bb56b75a186d0860b2b48b141a2fe08f020eefe252cdeb1d10ce52987046621ea5e05e683d711d71bb3494024fb08b10f4b08f10460ccfc8ff008620cb47ffbc7fc560083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010afbb549ca47856ff86e603af35a1fe2108f10460ccfc8ff0089d70de7067feab910083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

jonatack commented on src/txdb.cpp:
```diff
@@ -168,9 +168,34 @@ bool CBlockTreeDB::ReadLastBlockFile(int &nFile) {
+public:
+    // Prefer using CCoinsViewDB::Cursor() since we want to perform some
+    // cache warmup on instantiation.
+    CCoinsViewDBCursor(CDBIterator* pcursorIn, const uint256&hashBlockIn):
```
I'd suggest doing the formatting change in the first commit and making the second commit purely move-only.
```suggestion
    CCoinsViewDBCursor(CDBIterator* pcursorIn, const uint256& hashBlockIn):
```

jamesob replied:
Agree in principle but this commit is so easy to verify that I don't think it makes sense to rebase.

jonatack reviewed (commented):
re-ACK 7ad414f4bfa74595ee5726e66f3527045c02a977 modulo suggestion

jonatack commented:
> These are basically just the two recommendations from https://abseil.io/tips/134#recommendations

Thanks for the informative link @ryanofsky!

ryanofsky reviewed (approved):
Code review ACK 7ad414f4bfa74595ee5726e66f3527045c02a977. Two new commits look good and thanks for clarifying constructor comment
